src/Network/Receive.pm

```diff
@@ -266,18 +266,32 @@ sub parse_account_server_info {
 	} else {
 		delete $args->{lastLoginIP};
 	}
-
-	@{$args->{servers}} = map {
+	if ($args->{switch} eq '0069') {
+		@{$args->{servers}} = map {
```

Comment:
Looks like this would work, but there's a lot of code duplication. The only thing that's different is the unpack string, right?

Maybe we could do this instead:
```
my $unpack_str = 'a4 v Z20 v2 C2';
my @unpack_key = qw(ip port name users display);
if ($args->{switch} eq '0AC4') {
	$unpack_str = 'a4 v Z20 v3 a128';
	@unpack_key = qw(ip port name users state property unknown);
}
my $unpack_len = length pack $unpack_str;
@{$args->{servers}} = map {
	my %server;
	@server{@unpack_key} = unpack $unpack_str, $_;
	if ($masterServer && $masterServer->{private}) {
 		$server{ip} = $masterServer->{ip};
 	} else {
 		$server{ip} = inet_ntoa($server{ip});
 	}
 	$server{name} = bytesToString($server{name});
 	\%server
} unpack "(a$unpack_len)*", $args->{serverInfo};
```

What do you think?

Reply:
```
sub parse_account_server_info {
	my ($self, $args) = @_;

	if (length $args->{lastLoginIP} == 4 && $args->{lastLoginIP} ne "\0"x4) {
		$args->{lastLoginIP} = inet_ntoa($args->{lastLoginIP});
	} else {
		delete $args->{lastLoginIP};
	}
	if ($args->{switch} eq '0069') {
		my $unpack_str = 'a4 v Z20 v2 C2';
		my @unpack_key = qw(ip port name users display);
		$unpack_len = a32
	if ($args->{switch} eq '0AC4') {
		$unpack_str = 'a4 v Z20 v3 a128';
		@unpack_key = qw(ip port name users state property unknown);
		$unpack_len = a160
		}
		my $unpack_len = length pack $unpack_str;
		@{$args->{servers}} = map {
		my %server;
		@server{@unpack_key} = unpack $unpack_str, $_;
			if ($masterServer && $masterServer->{private}) {
			$server{ip} = $masterServer->{ip};
			} else {
			$server{ip} = inet_ntoa($server{ip});
			}
			$server{name} = bytesToString($server{name});
			\%server
		} unpack "(a$unpack_len)*", $args->{serverInfo};
	}
}	
```
my $unpack_len = length pack $unpack_str;
idk xD @allanon

Reply:
The idea is that we want to split the packet up into pieces which are exactly the number of bytes which the `$unpack_str` will consume. That's 32 bytes and 160 bytes respectively. We _could_ hardcode those two numbers... or we could calculate them. We can calculate them by giving `$unpack_str` to `pack`, which will create an output filled with null bytes, which is exactly the right number of bytes. We then use `length` to figure out how long that string is.

Your copy of the code won't work for a few reasons. The indentation is wrong, and if you fix it, it'll probably be more obvious why the code won't work.

Try this instead:

```Perl
sub parse_account_server_info {
	my ($self, $args) = @_;

	if (length $args->{lastLoginIP} == 4 && $args->{lastLoginIP} ne "\0"x4) {
		$args->{lastLoginIP} = inet_ntoa($args->{lastLoginIP});
	} else {
		delete $args->{lastLoginIP};
	}
	my $unpack_str = 'a4 v Z20 v2 C2';
	my @unpack_key = qw(ip port name users display);
	if ($args->{switch} eq '0AC4') {
		$unpack_str = 'a4 v Z20 v3 a128';
		@unpack_key = qw(ip port name users state property unknown);
	}
	my $unpack_len = length pack $unpack_str;
	@{$args->{servers}} = map {
		my %server;
		@server{@unpack_key} = unpack $unpack_str, $_;
		if ($masterServer && $masterServer->{private}) {
			$server{ip} = $masterServer->{ip};
		} else {
			$server{ip} = inet_ntoa($server{ip});
		}
		$server{name} = bytesToString($server{name});
		\%server
	} unpack "(a$unpack_len)*", $args->{serverInfo};
}
```

Reply:
so good but I'm not ok with `my $unpack_len = length pack $unpack_str;`
Maybe
```
	my $unpack_str = 'a4 v Z20 v2 C2';
	my @unpack_key = qw(ip port name users display);
	my $unpack_len = a32
	if ($args->{switch} eq '0AC4') {
		$unpack_str = 'a4 v Z20 v3 a128';
		@unpack_key = qw(ip port name users state property unknown);
        $unpack_len = a160
	}
```

Reply:
you can use something like this?
https://github.com/OpenKore/openkore/pull/789/files?diff=split

Reply:
so good , next day try it .